Race condition in terraformPluginSDKExternal causing provider restarts

### What happened?
This has been previously reported [here](https://github.com/crossplane-contrib/provider-upjet-aws/issues/1487) and [here](https://github.com/crossplane-contrib/provider-upjet-aws/issues/1426). Unfortunately, it hasn't been resolved yet.

This issue is prevalent in big environments where the provider manages many resources. The end result is that the Upjet provider pod restarts and affects stability in large Crossplane-managed environments using Upjet generated providers.

### How can we reproduce it?
This is observed in a large Upjet-managed environment, for example one with a few hundred SQS Queues.

### Root cause

I looked into this a bit and ran a race detector on Upjet itself, it fails with:

```
==================
WARNING: DATA RACE
Write at 0x000103c39580 by goroutine 82:
  github.com/crossplane/upjet/pkg/resource/fake.(*Observable).SetObservation()
      /Users/jciolek@alpha-sense.com/25feb/upjet/pkg/resource/fake/terraformed.go:32 +0x3c
  github.com/crossplane/upjet/pkg/resource/fake.(*Terraformed).SetObservation()
      <autogenerated>:1 +0x20
  github.com/crossplane/upjet/pkg/controller.(*terraformPluginSDKExternal).Update()
      /Users/jciolek@alpha-sense.com/25feb/upjet/pkg/controller/external_tfpluginsdk.go:718 +0x530
  github.com/crossplane/upjet/pkg/controller.(*terraformPluginSDKAsyncExternal).Update.func1()
      /Users/jciolek@alpha-sense.com/25feb/upjet/pkg/controller/external_async_tfpluginsdk.go:199 +0x2a0

Previous write at 0x000103c39580 by goroutine 79:
  github.com/crossplane/upjet/pkg/resource/fake.(*Observable).SetObservation()
      /Users/jciolek@alpha-sense.com/25feb/upjet/pkg/resource/fake/terraformed.go:32 +0x3c
  github.com/crossplane/upjet/pkg/resource/fake.(*Terraformed).SetObservation()
      <autogenerated>:1 +0x20
  github.com/crossplane/upjet/pkg/controller.(*terraformPluginSDKExternal).Create()
      /Users/jciolek@alpha-sense.com/25feb/upjet/pkg/controller/external_tfpluginsdk.go:660 +0x5c0
  github.com/crossplane/upjet/pkg/controller.(*terraformPluginSDKAsyncExternal).Create.func1()
      /Users/jciolek@alpha-sense.com/25feb/upjet/pkg/controller/external_async_tfpluginsdk.go:166 +0x2a0

Goroutine 82 (running) created at:
  github.com/crossplane/upjet/pkg/controller.(*terraformPluginSDKAsyncExternal).Update()
      /Users/jciolek@alpha-sense.com/25feb/upjet/pkg/controller/external_async_tfpluginsdk.go:178 +0x244
  github.com/crossplane/upjet/pkg/controller.TestAsyncTerraformPluginSDKUpdate.func3()
      /Users/jciolek@alpha-sense.com/25feb/upjet/pkg/controller/external_async_tfpluginsdk_test.go:286 +0x94
  testing.tRunner()
      /opt/homebrew/Cellar/go/1.24.0/libexec/src/testing/testing.go:1792 +0x180
  testing.(*T).Run.gowrap1()
      /opt/homebrew/Cellar/go/1.24.0/libexec/src/testing/testing.go:1851 +0x40

Goroutine 79 (finished) created at:
  github.com/crossplane/upjet/pkg/controller.(*terraformPluginSDKAsyncExternal).Create()
      /Users/jciolek@alpha-sense.com/25feb/upjet/pkg/controller/external_async_tfpluginsdk.go:145 +0x244
  github.com/crossplane/upjet/pkg/controller.TestAsyncTerraformPluginSDKCreate.func3()
      /Users/jciolek@alpha-sense.com/25feb/upjet/pkg/controller/external_async_tfpluginsdk_test.go:242 +0x94
  testing.tRunner()
      /opt/homebrew/Cellar/go/1.24.0/libexec/src/testing/testing.go:1792 +0x180
  testing.(*T).Run.gowrap1()
      /opt/homebrew/Cellar/go/1.24.0/libexec/src/testing/testing.go:1851 +0x40
==================
--- FAIL: TestConnect (0.00s)
    testing.go:1490: race detected during execution of test
```

It looks like it might be the cause of the problem, where the controller's Create and Update call the SetObservation method concurrently on a vanilla Go map which is not thread safe.

## Solution

Introduce synchronization or use a thread-safe data structure like [sync.Map](https://pkg.go.dev/sync#Map).

We've seen another failure related to this code, but now it's about writing to an uninitialized map. This occured in the S3 provider.

See:

```
2025/02/27 04:40:16 [DEBUG] Waiting for state to become: [success]
panic: assignment to entry in nil map

goroutine 292330486 [running]:
reflect.mapassign0(0xeb14160, 0xc002388020?, 0xc008c7dc38?, 0xc01880bd40?)
	runtime/map.go:1370 +0x1d
reflect.mapassign(0xc01880bd40?, 0xc005853b88?, 0xc005853bd8?, 0x40ffa9?)
	reflect/value.go:3828 +0x8d
github.com/modern-go/reflect2.(*UnsafeMapType).UnsafeSetIndex(...)
	github.com/modern-go/reflect2@v1.0.2/unsafe_map.go:76
github.com/json-iterator/go.(*mapDecoder).Decode(0xc006a17180, 0xc00f5114a0, 0xc01880bd40)
	github.com/json-iterator/go@v1.1.12/reflect_map.go:191 +0x325
github.com/json-iterator/go.(*placeholderDecoder).Decode(0xc00c557f00?, 0x4159eb?, 0xc005853c60?)
	github.com/json-iterator/go@v1.1.12/reflect.go:324 +0x1b
github.com/json-iterator/go.(*structFieldDecoder).Decode(0xc00b583640, 0xc00dba2000?, 0xc01880bd40)
	github.com/json-iterator/go@v1.1.12/reflect_struct_decoder.go:1054 +0x50
github.com/json-iterator/go.(*generalStructDecoder).decodeOneField(0xc00b583a80, 0xc004e4ced0?, 0xc01880bd40)
	github.com/json-iterator/go@v1.1.12/reflect_struct_decoder.go:552 +0x29c
github.com/json-iterator/go.(*generalStructDecoder).Decode(0xc00b583a80, 0xdb9d4e0?, 0xc01880bd40)
	github.com/json-iterator/go@v1.1.12/reflect_struct_decoder.go:508 +0x8b
github.com/json-iterator/go.(*Iterator).ReadVal(0xc01880bd40, {0xf00b7c0, 0xc00f511398})
	github.com/json-iterator/go@v1.1.12/reflect.go:79 +0x123
github.com/json-iterator/go.(*frozenConfig).Unmarshal(0xc0002394a0, {0xc00b42c700?, 0xc0173eeb70?, 0x10cb0ae0?}, {0xf00b7c0, 0xc00f511398})
	github.com/json-iterator/go@v1.1.12/config.go:348 +0x99
github.com/upbound/provider-aws/apis/s3/v1beta1.(*Bucket).SetObservation(0xc00f511200, 0xc00e725750?)
	github.com/upbound/provider-aws/apis/s3/v1beta1/zz_generated_terraformed.go:47 +0x7e
github.com/crossplane/upjet/pkg/controller.(*noForkExternal).Update(0xc018819040, {0x12df5968, 0xc010945f80}, {0x12e4c5c0?, 0xc00f511200})
	github.com/crossplane/upjet@v1.1.0-rc.0.0.20231227120826-4cb45f9104ac/pkg/controller/external_nofork.go:673 +0x27b
github.com/crossplane/upjet/pkg/controller.(*noForkAsyncExternal).Update.func1()
	github.com/crossplane/upjet@v1.1.0-rc.0.0.20231227120826-4cb45f9104ac/pkg/controller/external_async_nofork.go:161 +0x145
created by github.com/crossplane/upjet/pkg/controller.(*noForkAsyncExternal).Update in goroutine 3936
	github.com/crossplane/upjet@v1.1.0-rc.0.0.20231227120826-4cb45f9104ac/pkg/controller/external_async_nofork.go:157 +0x168
```

--- a/.github/workflows/ci.yml
+++ b/.github/workflows/ci.yml
@@ -196,3 +196,56 @@ jobs:
           flags: unittests
           files: _output/tests/linux_amd64/coverage.txt
           token: ${{ secrets.CODECOV_TOKEN }}
+
+  race-tests:
+    runs-on: ubuntu-latest
+    needs: detect-noop
+    if: needs.detect-noop.outputs.noop != 'true'
+    steps:
+      - name: Cleanup Disk
+        uses: jlumbroso/free-disk-space@54081f138730dfa15788a46383842cd2f914a1be # v1.3.1
+        with:
+          large-packages: false
+          swap-storage: false
+
+      - name: Checkout
+        uses: actions/checkout@de0fac2e4500dabe0009e67214ff5f5447ce83dd # v6.0.2
+        with:
+          submodules: true
+
+      - name: Fetch History
+        run: git fetch --prune --unshallow
+
+      - name: Setup Go
+        uses: actions/setup-go@4a3601121dd01d1626a1e23e37211e3254c1c06c # v6.4.0
+        with:
+          go-version: ${{ env.GO_VERSION }}
+
+      - name: Find the Go Build Cache
+        id: go-cache-paths
+        run: |
+          echo "go-build=$(make go.cachedir)" >> $GITHUB_OUTPUT
+          echo "go-mod=$(make go.mod.cachedir)" >> $GITHUB_OUTPUT
+
+      - name: Cache the Go Build Cache
+        uses: actions/cache@27d5ce7f107fe9357f9df03efb73ab90386fccae # v5.0.5
+        with:
+          path: ${{ steps.go-cache-paths.outputs.go-build }}
+          key: ${{ runner.os }}-build-race-tests-${{ hashFiles('**/go.sum') }}
+          restore-keys: ${{ runner.os }}-build-race-tests-
+
+      - name: Cache Go Dependencies
+        uses: actions/cache@27d5ce7f107fe9357f9df03efb73ab90386fccae # v5.0.5
+        with:
+          path: ${{ steps.go-cache-paths.outputs.go-mod }}
+          key: ${{ runner.os }}-pkg-race-tests-${{ hashFiles('**/go.sum') }}
+          restore-keys: ${{ runner.os }}-pkg-race-tests-
+
+      - name: Vendor Dependencies
+        run: make vendor vendor.check
+
+      # Runs the unit tests with the Go race detector enabled to guard against
+      # regressions of concurrency bugs such as the managed resource status
+      # update data race. See: https://github.com/crossplane/upjet/issues/472
+      - name: Run Race Tests
+        run: make test.race

--- a/Makefile
+++ b/Makefile
@@ -70,4 +70,14 @@ go.cachedir:
 go.mod.cachedir:
 	@go env GOMODCACHE
 
-.PHONY: reviewable submodules fallthrough go.mod.cachedir go.cachedir
+# test.race runs the unit tests with the Go race detector enabled. The race
+# detector requires cgo, so we enable it here.
+# Do NOT directly use the `go.test.unit` make target of
+# the build submodule here. There's a `pipefail` issue with that target,
+# which will mask any failing tests.
+test.race:
+	@$(INFO) go test race
+	@CGO_ENABLED=1 $(GOHOST) test -race $(GO_STATIC_FLAGS) $(GO_PACKAGES) || $(FAIL)
+	@$(OK) go test race
+
+.PHONY: reviewable submodules fallthrough go.mod.cachedir go.cachedir test.race

--- a/pkg/controller/external_async_tfpluginfw.go
+++ b/pkg/controller/external_async_tfpluginfw.go
@@ -209,6 +209,10 @@ func (n *terraformPluginFrameworkAsyncExternalClient) Create(_ context.Context, 
 	}
 
 	ctx, cancel := context.WithDeadline(context.Background(), n.opTracker.LastOperation.StartTime().Add(defaultAsyncTimeout))
+	// We deep-copy the managed resource to prevent a data race between the
+	// goroutine we are about to start below and the managed reconciler.
+	// Please see: https://github.com/crossplane/upjet/issues/472
+	mgCopy := mg.DeepCopyObject().(xpresource.Managed)
 	go func() {
 		// The order of deferred functions, executed last-in-first-out, is
 		// significant. The context should be canceled last, because it is
@@ -225,8 +229,8 @@ func (n *terraformPluginFrameworkAsyncExternalClient) Create(_ context.Context, 
 
 			n.opTracker.LastOperation.MarkEnd()
 			name := types.NamespacedName{
-				Namespace: mg.GetNamespace(),
-				Name:      mg.GetName(),
+				Namespace: mgCopy.GetNamespace(),
+				Name:      mgCopy.GetName(),
 			}
 			// we request an immediate reconcile upon success to set the status, or
 			// in case of failure (err != nil), if there's no cached error.
@@ -239,7 +243,7 @@ func (n *terraformPluginFrameworkAsyncExternalClient) Create(_ context.Context, 
 		defer ph.recoverIfPanic(ctx)
 
 		n.opTracker.logger.Debug("Async create starting...")
-		_, ph.err = n.terraformPluginFrameworkExternalClient.Create(ctx, mg)
+		_, ph.err = n.terraformPluginFrameworkExternalClient.Create(ctx, mgCopy)
 	}()
 
 	return managed.ExternalCreation{}, n.opTracker.LastOperation.Error()
@@ -251,6 +255,10 @@ func (n *terraformPluginFrameworkAsyncExternalClient) Update(_ context.Context, 
 	}
 
 	ctx, cancel := context.WithDeadline(context.Background(), n.opTracker.LastOperation.StartTime().Add(defaultAsyncTimeout))
+	// We deep-copy the managed resource to prevent a data race between the
+	// goroutine we are about to start below and the managed reconciler.
+	// Please see: https://github.com/crossplane/upjet/issues/472
+	mgCopy := mg.DeepCopyObject().(xpresource.Managed)
 	go func() {
 		// The order of deferred functions, executed last-in-first-out, is
 		// significant. The context should be canceled last, because it is
@@ -267,8 +275,8 @@ func (n *terraformPluginFrameworkAsyncExternalClient) Update(_ context.Context, 
 
 			n.opTracker.LastOperation.MarkEnd()
 			name := types.NamespacedName{
-				Namespace: mg.GetNamespace(),
-				Name:      mg.GetName(),
+				Namespace: mgCopy.GetNamespace(),
+				Name:      mgCopy.GetName(),
 			}
 			// we request an immediate reconcile upon success to set the status, or
 			// in case of failure (err != nil), if there's no cached error.
@@ -281,7 +289,7 @@ func (n *terraformPluginFrameworkAsyncExternalClient) Update(_ context.Context, 
 		defer ph.recoverIfPanic(ctx)
 
 		n.opTracker.logger.Debug("Async update starting...")
-		_, ph.err = n.terraformPluginFrameworkExternalClient.Update(ctx, mg)
+		_, ph.err = n.terraformPluginFrameworkExternalClient.Update(ctx, mgCopy)
 	}()
 
 	return managed.ExternalUpdate{}, n.opTracker.LastOperation.Error()

--- a/pkg/controller/external_async_tfpluginfw_test.go
+++ b/pkg/controller/external_async_tfpluginfw_test.go
@@ -1,0 +1,436 @@
+// SPDX-FileCopyrightText: 2026 The Crossplane Authors <https://crossplane.io>
+//
+// SPDX-License-Identifier: Apache-2.0
+
+package controller
+
+import (
+	"context"
+	"testing"
+
+	"github.com/crossplane/crossplane-runtime/v2/pkg/reconciler/managed"
+	xpresource "github.com/crossplane/crossplane-runtime/v2/pkg/resource"
+	"github.com/crossplane/crossplane-runtime/v2/pkg/test"
+	"github.com/google/go-cmp/cmp"
+	"k8s.io/apimachinery/pkg/types"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+
+	"github.com/crossplane/upjet/v2/pkg/config"
+	"github.com/crossplane/upjet/v2/pkg/terraform"
+)
+
+func prepareTerraformPluginFrameworkAsyncExternal(testConfig testConfiguration, fns CallbackFns) *terraformPluginFrameworkAsyncExternalClient {
+	return &terraformPluginFrameworkAsyncExternalClient{
+		terraformPluginFrameworkExternalClient: prepareTPFExternalWithTestConfig(testConfig),
+		callback:                               fns,
+	}
+}
+
+func TestAsyncTerraformPluginFrameworkConnect(t *testing.T) {
+	type args struct {
+		setupFn terraform.SetupFn
+		cfg     *config.Resource
+		ots     *OperationTrackerStore
+		obj     xpresource.Managed
+	}
+	type want struct {
+		err error
+	}
+	cases := map[string]struct {
+		args
+		want
+	}{
+		"Successful": {
+			args: args{
+				setupFn: func(_ context.Context, _ client.Client, _ xpresource.Managed) (terraform.Setup, error) {
+					return terraform.Setup{
+						FrameworkProvider: &mockTPFProvider{},
+					}, nil
+				},
+				cfg: newBaseUpjetConfig(),
+				obj: newObjAsync(),
+				ots: ots,
+			},
+		},
+	}
+	for name, tc := range cases {
+		t.Run(name, func(t *testing.T) {
+			c := NewTerraformPluginFrameworkAsyncConnector(nil, tc.args.ots, tc.args.setupFn, tc.args.cfg, WithTerraformPluginFrameworkAsyncLogger(logTest))
+			_, err := c.Connect(t.Context(), tc.args.obj)
+			if diff := cmp.Diff(tc.want.err, err, test.EquateErrors()); diff != "" {
+				t.Errorf("\n%s\nConnect(...): -want error, +got error:\n", diff)
+			}
+		})
+	}
+}
+
+func TestAsyncTerraformPluginFrameworkObserve(t *testing.T) {
+	type want struct {
+		obs managed.ExternalObservation
+		err error
+	}
+	cases := map[string]struct {
+		testConfiguration
+		want
+	}{
+		"NotExists": {
+			testConfiguration: testConfiguration{
+				r:               newMockBaseTPFResource(),
+				cfg:             newBaseUpjetConfig(),
+				obj:             newBaseObject(),
+				currentStateMap: nil,
+				plannedStateMap: map[string]any{
+					"name": "example",
+				},
+				params: map[string]any{
+					"name": "example",
+				},
+			},
+			want: want{
+				obs: managed.ExternalObservation{
+					ResourceExists:          false,
+					ResourceUpToDate:        false,
+					ResourceLateInitialized: false,
+					ConnectionDetails:       nil,
+					Diff:                    "",
+				},
+			},
+		},
+		"UpToDate": {
+			testConfiguration: testConfiguration{
+				r:   newMockBaseTPFResource(),
+				cfg: newBaseUpjetConfig(),
+				obj: newBaseObject(),
+				params: map[string]any{
+					"id":   "example-id",
+					"name": "example",
+				},
+				currentStateMap: map[string]any{
+					"id":   "example-id",
+					"name": "example",
+				},
+				plannedStateMap: map[string]any{
+					"id":   "example-id",
+					"name": "example",
+				},
+			},
+			want: want{
+				obs: managed.ExternalObservation{
+					ResourceExists:          true,
+					ResourceUpToDate:        true,
+					ResourceLateInitialized: true,
+					ConnectionDetails:       nil,
+					Diff:                    "",
+				},
+			},
+		},
+	}
+	for name, tc := range cases {
+		t.Run(name, func(t *testing.T) {
+			terraformPluginFrameworkAsyncExternal := prepareTerraformPluginFrameworkAsyncExternal(tc.testConfiguration, CallbackFns{})
+			observation, err := terraformPluginFrameworkAsyncExternal.Observe(t.Context(), &tc.testConfiguration.obj)
+			if diff := cmp.Diff(tc.want.obs, observation); diff != "" {
+				t.Errorf("\n%s\nObserve(...): -want observation, +got observation:\n", diff)
+			}
+			if diff := cmp.Diff(tc.want.err, err, test.EquateErrors()); diff != "" {
+				t.Errorf("\n%s\nConnect(...): -want error, +got error:\n", diff)
+			}
+		})
+	}
+}
+
+func TestAsyncTerraformPluginFrameworkCreate(t *testing.T) {
+	type want struct {
+		err error
+	}
+	cases := map[string]struct {
+		testConfiguration
+		want
+	}{
+		"Successful": {
+			testConfiguration: testConfiguration{
+				r:               newMockBaseTPFResource(),
+				cfg:             newBaseUpjetConfig(),
+				obj:             newBaseObject(),
+				currentStateMap: nil,
+				plannedStateMap: map[string]any{
+					"name": "example",
+				},
+				params: map[string]any{
+					"name": "example",
+				},
+				newStateMap: map[string]any{
+					"name": "example",
+					"id":   "example-id",
+				},
+			},
+		},
+	}
+	for name, tc := range cases {
+		t.Run(name, func(t *testing.T) {
+			terraformPluginFrameworkAsyncExternal := prepareTerraformPluginFrameworkAsyncExternal(tc.testConfiguration, CallbackFns{
+				CreateFn: func(_ types.NamespacedName) terraform.CallbackFn {
+					return func(err error, ctx context.Context) error {
+						return nil
+					}
+				},
+			})
+			_, err := terraformPluginFrameworkAsyncExternal.Create(t.Context(), &tc.testConfiguration.obj)
+			if diff := cmp.Diff(tc.want.err, err, test.EquateErrors()); diff != "" {
+				t.Errorf("\n%s\nterraformPluginFrameworkAsyncExternalClient.Create(...): -want error, +got error:\n", diff)
+			}
+		})
+	}
+}
+
+func TestAsyncTerraformPluginFrameworkUpdate(t *testing.T) {
+	type want struct {
+		err error
+	}
+	cases := map[string]struct {
+		testConfiguration
+		want
+	}{
+		"Successful": {
+			testConfiguration: testConfiguration{
+				r:   newMockBaseTPFResource(),
+				cfg: newBaseUpjetConfig(),
+				obj: newBaseObject(),
+				currentStateMap: map[string]any{
+					"name": "example",
+					"id":   "example-id",
+				},
+				plannedStateMap: map[string]any{
+					"name": "example-updated",
+					"id":   "example-id",
+				},
+				params: map[string]any{
+					"name": "example-updated",
+				},
+				newStateMap: map[string]any{
+					"name": "example-updated",
+					"id":   "example-id",
+				},
+			},
+		},
+	}
+	for name, tc := range cases {
+		t.Run(name, func(t *testing.T) {
+			terraformPluginFrameworkAsyncExternal := prepareTerraformPluginFrameworkAsyncExternal(tc.testConfiguration, CallbackFns{
+				UpdateFn: func(_ types.NamespacedName) terraform.CallbackFn {
+					return func(err error, ctx context.Context) error {
+						return nil
+					}
+				},
+			})
+			_, err := terraformPluginFrameworkAsyncExternal.Update(t.Context(), &tc.testConfiguration.obj)
+			if diff := cmp.Diff(tc.want.err, err, test.EquateErrors()); diff != "" {
+				t.Errorf("\n%s\nterraformPluginFrameworkAsyncExternalClient.Update(...): -want error, +got error:\n", diff)
+			}
+		})
+	}
+}
+
+func TestAsyncTerraformPluginFrameworkDelete(t *testing.T) {
+	type want struct {
+		err error
+	}
+	cases := map[string]struct {
+		testConfiguration
+		want
+	}{
+		"Successful": {
+			testConfiguration: testConfiguration{
+				r:   newMockBaseTPFResource(),
+				cfg: newBaseUpjetConfig(),
+				obj: newBaseObject(),
+				currentStateMap: map[string]any{
+					"name": "example",
+					"id":   "example-id",
+				},
+				plannedStateMap: nil,
+				params: map[string]any{
+					"name": "example",
+				},
+				newStateMap: nil,
+			},
+		},
+	}
+	for name, tc := range cases {
+		t.Run(name, func(t *testing.T) {
+			terraformPluginFrameworkAsyncExternal := prepareTerraformPluginFrameworkAsyncExternal(tc.testConfiguration, CallbackFns{
+				DestroyFn: func(_ types.NamespacedName) terraform.CallbackFn {
+					return func(err error, ctx context.Context) error {
+						return nil
+					}
+				},
+			})
+			_, err := terraformPluginFrameworkAsyncExternal.Delete(t.Context(), &tc.testConfiguration.obj)
+			if diff := cmp.Diff(tc.want.err, err, test.EquateErrors()); diff != "" {
+				t.Errorf("\n%s\nterraformPluginFrameworkAsyncExternalClient.Delete(...): -want error, +got error:\n", diff)
+			}
+		})
+	}
+}
+
+// TestAsyncTerraformPluginFrameworkCreateRace is a regression test for
+// the data race on a managed resource's status, between upjet's async Create
+// operation and the managed reconciler.
+// Must be run with `go test -race`.
+//
+// Please also see: https://github.com/crossplane/upjet/issues/472.
+func TestAsyncTerraformPluginFrameworkCreateRace(t *testing.T) {
+	obj := newObjAsync()
+	tc := testConfiguration{
+		r:               newMockBaseTPFResource(),
+		cfg:             newBaseUpjetConfig(),
+		currentStateMap: nil,
+		plannedStateMap: map[string]any{
+			"name": "example",
+		},
+		params: map[string]any{
+			"name": "example",
+		},
+		newStateMap: map[string]any{
+			"name": "example",
+			"id":   "example-id",
+		},
+	}
+
+	extDone := make(chan struct{})
+	ext := prepareTerraformPluginFrameworkAsyncExternal(tc, CallbackFns{
+		CreateFn: func(_ types.NamespacedName) terraform.CallbackFn {
+			return func(_ error, _ context.Context) error {
+				// Signal the async operation of the external client has completed.
+				close(extDone)
+				return nil
+			}
+		},
+	})
+	// This call starts the async worker that will race with
+	// the managed reconciler below.
+	if _, err := ext.Create(t.Context(), obj); err != nil {
+		t.Fatalf("terraformPluginFrameworkAsyncExternalClient.Create(...): unexpected error: %v", err)
+	}
+
+	// Simulate the managed reconciler concurrently writing to the status of
+	// the same MR (obj above).
+	mrDone := make(chan struct{})
+	go func() {
+		_ = obj.DeepCopyObject()
+		_ = obj.SetObservation(map[string]any{"name": "example"})
+		// Signal the managed reconciler has completed.
+		close(mrDone)
+	}()
+	<-extDone
+	<-mrDone
+}
+
+// TestAsyncTerraformPluginFrameworkUpdateRace is a regression test for
+// the data race on a managed resource's status, between upjet's async Update
+// operation and the managed reconciler.
+// Must be run with `go test -race`.
+//
+// Please also see: https://github.com/crossplane/upjet/issues/472.
+func TestAsyncTerraformPluginFrameworkUpdateRace(t *testing.T) {
+	obj := newObjAsync()
+	tc := testConfiguration{
+		r:   newMockBaseTPFResource(),
+		cfg: newBaseUpjetConfig(),
+		currentStateMap: map[string]any{
+			"name": "example",
+			"id":   "example-id",
+		},
+		plannedStateMap: map[string]any{
+			"name": "example-updated",
+			"id":   "example-id",
+		},
+		params: map[string]any{
+			"name": "example-updated",
+		},
+		newStateMap: map[string]any{
+			"name": "example-updated",
+			"id":   "example-id",
+		},
+	}
+
+	extDone := make(chan struct{})
+	ext := prepareTerraformPluginFrameworkAsyncExternal(tc, CallbackFns{
+		UpdateFn: func(_ types.NamespacedName) terraform.CallbackFn {
+			return func(_ error, _ context.Context) error {
+				// Signal the async operation of the external client has completed.
+				close(extDone)
+				return nil
+			}
+		},
+	})
+	// This call starts the async worker that will race with
+	// the managed reconciler below.
+	if _, err := ext.Update(t.Context(), obj); err != nil {
+		t.Fatalf("terraformPluginFrameworkAsyncExternalClient.Update(...): unexpected error: %v", err)
+	}
+
+	// Simulate the managed reconciler concurrently writing to the status of
+	// the same MR (obj above).
+	mrDone := make(chan struct{})
+	go func() {
+		_ = obj.DeepCopyObject()
+		_ = obj.SetObservation(map[string]any{"name": "example"})
+		// Signal the managed reconciler has completed.
+		close(mrDone)
+	}()
+	<-extDone
+	<-mrDone
+}
+
+// TestAsyncTerraformPluginFrameworkDeleteRace is a guard test asserting that
+// upjet's async Delete operation does not concurrently access a managed
+// resource's status while the managed reconciler does.
+// Current async client Delete does not modify MR status or spec.
+// Must be run with `go test -race`.
+func TestAsyncTerraformPluginFrameworkDeleteRace(t *testing.T) {
+	obj := newObjAsync()
+	tc := testConfiguration{
+		r:   newMockBaseTPFResource(),
+		cfg: newBaseUpjetConfig(),
+		currentStateMap: map[string]any{
+			"name": "example",
+			"id":   "example-id",
+		},
+		plannedStateMap: nil,
+		params: map[string]any{
+			"name": "example",
+		},
+		newStateMap: nil,
+	}
+
+	extDone := make(chan struct{})
+	ext := prepareTerraformPluginFrameworkAsyncExternal(tc, CallbackFns{
+		DestroyFn: func(_ types.NamespacedName) terraform.CallbackFn {
+			return func(_ error, _ context.Context) error {
+				// Signal the async operation of the external client has completed.
+				close(extDone)
+				return nil
+			}
+		},
+	})
+	// This call starts the async worker that will race with
+	// the managed reconciler below.
+	if _, err := ext.Delete(t.Context(), obj); err != nil {
+		t.Fatalf("terraformPluginFrameworkAsyncExternalClient.Delete(...): unexpected error: %v", err)
+	}
+
+	// Simulate the managed reconciler concurrently writing to the status of
+	// the same MR (obj above).
+	mrDone := make(chan struct{})
+	go func() {
+		_ = obj.DeepCopyObject()
+		// Managed reconciler does not call SetObservation during deletion.
+		// This is an extra check at the moment.
+		_ = obj.SetObservation(map[string]any{"name": "example"})
+		// Signal the managed reconciler has completed.
+		close(mrDone)
+	}()
+	<-extDone
+	<-mrDone
+}

--- a/pkg/controller/external_async_tfpluginsdk.go
+++ b/pkg/controller/external_async_tfpluginsdk.go
@@ -143,6 +143,10 @@ func (n *terraformPluginSDKAsyncExternal) Create(_ context.Context, mg xpresourc
 	}
 
 	ctx, cancel := context.WithDeadline(context.Background(), n.opTracker.LastOperation.StartTime().Add(defaultAsyncTimeout))
+	// We deep-copy the managed resource to prevent a data race between the
+	// goroutine we are about to start below and the managed reconciler.
+	// Please see: https://github.com/crossplane/upjet/issues/472
+	mgCopy := mg.DeepCopyObject().(xpresource.Managed)
 	go func() {
 		// The order of deferred functions, executed last-in-first-out, is
 		// significant. The context should be canceled last, because it is
@@ -159,8 +163,8 @@ func (n *terraformPluginSDKAsyncExternal) Create(_ context.Context, mg xpresourc
 
 			n.opTracker.LastOperation.MarkEnd()
 			name := types.NamespacedName{
-				Namespace: mg.GetNamespace(),
-				Name:      mg.GetName(),
+				Namespace: mgCopy.GetNamespace(),
+				Name:      mgCopy.GetName(),
 			}
 			// we request an immediate reconcile upon success to set the status, or
 			// in case of failure (err != nil), if there's no cached error.
@@ -173,7 +177,7 @@ func (n *terraformPluginSDKAsyncExternal) Create(_ context.Context, mg xpresourc
 		defer ph.recoverIfPanic(ctx)
 
 		n.opTracker.logger.Debug("Async create starting...", "tfID", n.opTracker.GetTfID())
-		_, ph.err = n.terraformPluginSDKExternal.Create(ctx, mg)
+		_, ph.err = n.terraformPluginSDKExternal.Create(ctx, mgCopy)
 	}()
 
 	return managed.ExternalCreation{}, n.opTracker.LastOperation.Error()
@@ -185,6 +189,10 @@ func (n *terraformPluginSDKAsyncExternal) Update(_ context.Context, mg xpresourc
 	}
 
 	ctx, cancel := context.WithDeadline(context.Background(), n.opTracker.LastOperation.StartTime().Add(defaultAsyncTimeout))
+	// We deep-copy the managed resource to prevent a data race between the
+	// goroutine we are about to start below and the managed reconciler.
+	// Please see: https://github.com/crossplane/upjet/issues/472
+	mgCopy := mg.DeepCopyObject().(xpresource.Managed)
 	go func() {
 		// The order of deferred functions, executed last-in-first-out, is
 		// significant. The context should be canceled last, because it is
@@ -201,8 +209,8 @@ func (n *terraformPluginSDKAsyncExternal) Update(_ context.Context, mg xpresourc
 
 			n.opTracker.LastOperation.MarkEnd()
 			name := types.NamespacedName{
-				Namespace: mg.GetNamespace(),
-				Name:      mg.GetName(),
+				Namespace: mgCopy.GetNamespace(),
+				Name:      mgCopy.GetName(),
 			}
 			if cErr := n.callback.Update(name, err == nil || currentErr == nil)(err, ctx); cErr != nil {
 				n.opTracker.logger.Info("Async update callback failed", "error", cErr.Error())
@@ -211,7 +219,7 @@ func (n *terraformPluginSDKAsyncExternal) Update(_ context.Context, mg xpresourc
 		defer ph.recoverIfPanic(ctx)
 
 		n.opTracker.logger.Debug("Async update starting...", "tfID", n.opTracker.GetTfID())
-		_, ph.err = n.terraformPluginSDKExternal.Update(ctx, mg)
+		_, ph.err = n.terraformPluginSDKExternal.Update(ctx, mgCopy)
 	}()
 
 	return managed.ExternalUpdate{}, n.opTracker.LastOperation.Error()

--- a/pkg/controller/external_async_tfpluginsdk.go
+++ b/pkg/controller/external_async_tfpluginsdk.go
@@ -212,6 +212,10 @@ func (n *terraformPluginSDKAsyncExternal) Update(_ context.Context, mg xpresourc
 				Namespace: mgCopy.GetNamespace(),
 				Name:      mgCopy.GetName(),
 			}
+			// we request an immediate reconcile upon success to set the status, or
+			// in case of failure (err != nil), if there's no cached error.
+			// If there already exists a cached error, managed reconciler
+			// will already requeue.
 			if cErr := n.callback.Update(name, err == nil || currentErr == nil)(err, ctx); cErr != nil {
 				n.opTracker.logger.Info("Async update callback failed", "error", cErr.Error())
 			}

--- a/pkg/controller/external_async_tfpluginsdk_test.go
+++ b/pkg/controller/external_async_tfpluginsdk_test.go
@@ -113,7 +113,7 @@ func TestAsyncTerraformPluginSDKConnect(t *testing.T) {
 	for name, tc := range cases {
 		t.Run(name, func(t *testing.T) {
 			c := NewTerraformPluginSDKAsyncConnector(nil, tc.args.ots, tc.args.setupFn, tc.args.cfg, WithTerraformPluginSDKAsyncLogger(logTest))
-			_, err := c.Connect(context.TODO(), tc.args.obj)
+			_, err := c.Connect(t.Context(), tc.args.obj)
 			if diff := cmp.Diff(tc.want.err, err, test.EquateErrors()); diff != "" {
 				t.Errorf("\n%s\nConnect(...): -want error, +got error:\n", diff)
 			}
@@ -179,7 +179,7 @@ func TestAsyncTerraformPluginSDKObserve(t *testing.T) {
 	for name, tc := range cases {
 		t.Run(name, func(t *testing.T) {
 			terraformPluginSDKAsyncExternal := prepareTerraformPluginSDKAsyncExternal(tc.args.r, tc.args.cfg, CallbackFns{})
-			observation, err := terraformPluginSDKAsyncExternal.Observe(context.TODO(), tc.args.obj)
+			observation, err := terraformPluginSDKAsyncExternal.Observe(t.Context(), tc.args.obj)
 			if diff := cmp.Diff(tc.want.obs, observation); diff != "" {
 				t.Errorf("\n%s\nObserve(...): -want observation, +got observation:\n", diff)
 			}
@@ -226,7 +226,7 @@ func TestAsyncTerraformPluginSDKCreate(t *testing.T) {
 	for name, tc := range cases {
 		t.Run(name, func(t *testing.T) {
 			terraformPluginSDKAsyncExternal := prepareTerraformPluginSDKAsyncExternal(tc.args.r, tc.args.cfg, tc.args.fns)
-			_, err := terraformPluginSDKAsyncExternal.Create(context.TODO(), tc.args.obj)
+			_, err := terraformPluginSDKAsyncExternal.Create(t.Context(), tc.args.obj)
 			if diff := cmp.Diff(tc.want.err, err, test.EquateErrors()); diff != "" {
 				t.Errorf("\n%s\nterraformPluginSDKAsyncExternal.Create(...): -want error, +got error:\n", diff)
 			}
@@ -270,7 +270,7 @@ func TestAsyncTerraformPluginSDKUpdate(t *testing.T) {
 	for name, tc := range cases {
 		t.Run(name, func(t *testing.T) {
 			terraformPluginSDKAsyncExternal := prepareTerraformPluginSDKAsyncExternal(tc.args.r, tc.args.cfg, tc.args.fns)
-			_, err := terraformPluginSDKAsyncExternal.Update(context.TODO(), tc.args.obj)
+			_, err := terraformPluginSDKAsyncExternal.Update(t.Context(), tc.args.obj)
 			if diff := cmp.Diff(tc.want.err, err, test.EquateErrors()); diff != "" {
 				t.Errorf("\n%s\nterraformPluginSDKAsyncExternal.Update(...): -want error, +got error:\n", diff)
 			}
@@ -314,12 +314,142 @@ func TestAsyncTerraformPluginSDKDelete(t *testing.T) {
 	for name, tc := range cases {
 		t.Run(name, func(t *testing.T) {
 			terraformPluginSDKAsyncExternal := prepareTerraformPluginSDKAsyncExternal(tc.args.r, tc.args.cfg, tc.args.fns)
-			_, err := terraformPluginSDKAsyncExternal.Delete(context.TODO(), tc.args.obj)
+			_, err := terraformPluginSDKAsyncExternal.Delete(t.Context(), tc.args.obj)
 			if diff := cmp.Diff(tc.want.err, err, test.EquateErrors()); diff != "" {
 				t.Errorf("\n%s\nterraformPluginSDKAsyncExternal.Delete(...): -want error, +got error:\n", diff)
 			}
 		})
 	}
+}
+
+// TestAsyncTerraformPluginSDKCreateRace is a regression test for
+// the data race on a managed resource's status, between upjet's async Create
+// operation and the managed reconciler.
+// Must be run with `go test -race`.
+//
+// Please also see: https://github.com/crossplane/upjet/issues/472.
+func TestAsyncTerraformPluginSDKCreateRace(t *testing.T) {
+	obj := newObjAsync()
+	r := mockResource{
+		ApplyFn: func(_ context.Context, _ *tf.InstanceState, _ *tf.InstanceDiff, _ interface{}) (*tf.InstanceState, diag.Diagnostics) {
+			return &tf.InstanceState{ID: "example-id", Attributes: map[string]string{"name": "example"}}, nil
+		},
+	}
+
+	extDone := make(chan struct{})
+	ext := prepareTerraformPluginSDKAsyncExternal(r, cfgAsync, CallbackFns{
+		CreateFn: func(_ types.NamespacedName) terraform.CallbackFn {
+			return func(_ error, _ context.Context) error {
+				// Signal the async operation of the external client has completed.
+				close(extDone)
+				return nil
+			}
+		},
+	})
+	// This call starts the async worker that will race with
+	// the managed reconciler below.
+	if _, err := ext.Create(t.Context(), obj); err != nil {
+		t.Fatalf("terraformPluginSDKAsyncExternal.Create(...): unexpected error: %v", err)
+	}
+
+	// Simulate the managed reconciler concurrently writing to the status of
+	// the same MR (obj above).
+	mrDone := make(chan struct{})
+	go func() {
+		_ = obj.DeepCopyObject()
+		_ = obj.SetObservation(map[string]any{"name": "example"})
+		// Signal the managed reconciler has completed.
+		close(mrDone)
+	}()
+	<-extDone
+	<-mrDone
+}
+
+// TestAsyncTerraformPluginSDKUpdateRace is a regression test for
+// the data race on a managed resource's status, between upjet's async Update
+// operation and the managed reconciler.
+// Must be run with `go test -race`.
+//
+// Please also see: https://github.com/crossplane/upjet/issues/472.
+func TestAsyncTerraformPluginSDKUpdateRace(t *testing.T) {
+	obj := newObjAsync()
+	r := mockResource{
+		ApplyFn: func(_ context.Context, _ *tf.InstanceState, _ *tf.InstanceDiff, _ interface{}) (*tf.InstanceState, diag.Diagnostics) {
+			return &tf.InstanceState{ID: "example-id", Attributes: map[string]string{"name": "example"}}, nil
+		},
+	}
+
+	extDone := make(chan struct{})
+	ext := prepareTerraformPluginSDKAsyncExternal(r, cfgAsync, CallbackFns{
+		UpdateFn: func(_ types.NamespacedName) terraform.CallbackFn {
+			return func(_ error, _ context.Context) error {
+				// Signal the async operation of the external client has completed.
+				close(extDone)
+				return nil
+			}
+		},
+	})
+	// This call starts the async worker that will race with
+	// the managed reconciler below.
+	if _, err := ext.Update(t.Context(), obj); err != nil {
+		t.Fatalf("terraformPluginSDKAsyncExternal.Update(...): unexpected error: %v", err)
+	}
+
+	// Simulate the managed reconciler concurrently writing to the status of
+	// the same MR (obj above).
+	mrDone := make(chan struct{})
+	go func() {
+		_ = obj.DeepCopyObject()
+		_ = obj.SetObservation(map[string]any{"name": "example"})
+		// Signal the managed reconciler has completed.
+		close(mrDone)
+	}()
+	<-extDone
+	<-mrDone
+}
+
+// TestAsyncTerraformPluginSDKDeleteRace is a guard test asserting that upjet's
+// async Delete operation does not concurrently access a managed resource's
+// status while the managed reconciler does. Current async client Delete
+// implementation does not modify MR status or spec.
+// Must be run with `go test -race`.
+func TestAsyncTerraformPluginSDKDeleteRace(t *testing.T) {
+	obj := newObjAsync()
+	r := mockResource{
+		ApplyFn: func(_ context.Context, _ *tf.InstanceState, _ *tf.InstanceDiff, _ interface{}) (*tf.InstanceState, diag.Diagnostics) {
+			return &tf.InstanceState{ID: "example-id", Attributes: map[string]string{"name": "example"}}, nil
+		},
+	}
+
+	extDone := make(chan struct{})
+	ext := prepareTerraformPluginSDKAsyncExternal(r, cfgAsync, CallbackFns{
+		DestroyFn: func(_ types.NamespacedName) terraform.CallbackFn {
+			return func(_ error, _ context.Context) error {
+				// Signal the async operation of the external client has completed.
+				close(extDone)
+				return nil
+			}
+		},
+	})
+	// This call starts the async worker that will race with
+	// the managed reconciler below.
+	if _, err := ext.Delete(t.Context(), obj); err != nil {
+		t.Fatalf("terraformPluginSDKAsyncExternal.Delete(...): unexpected error: %v", err)
+	}
+
+	// Simulate the managed reconciler concurrently writing to the status of
+	// the same MR (obj above).
+	mrDone := make(chan struct{})
+	go func() {
+		_ = obj.DeepCopyObject()
+		// Managed reconciler does not call SetObservation during deletion.
+		// This is an extra check at the moment.
+		_ = obj.SetObservation(map[string]any{"name": "example"})
+		// Signal the managed reconciler has completed.
+		close(mrDone)
+	}()
+	<-extDone
+	<-mrDone
 }
 
 func newObjAsync() *fake.Terraformed {

--- a/pkg/controller/external_async_tfpluginsdk_test.go
+++ b/pkg/controller/external_async_tfpluginsdk_test.go
@@ -61,20 +61,6 @@ var (
 			return nil, nil
 		}},
 	}
-	objAsync = &fake.Terraformed{
-		Parameterizable: fake.Parameterizable{
-			Parameters: map[string]any{
-				"name": "example",
-				"map": map[string]any{
-					"key": "value",
-				},
-				"list": []any{"elem1", "elem2"},
-			},
-		},
-		Observable: fake.Observable{
-			Observation: map[string]any{},
-		},
-	}
 )
 
 func prepareTerraformPluginSDKAsyncExternal(r Resource, cfg *config.Resource, fns CallbackFns) *terraformPluginSDKAsyncExternal {
@@ -119,7 +105,7 @@ func TestAsyncTerraformPluginSDKConnect(t *testing.T) {
 					return terraform.Setup{}, nil
 				},
 				cfg: cfgAsync,
-				obj: objAsync,
+				obj: newObjAsync(),
 				ots: ots,
 			},
 		},
@@ -157,7 +143,7 @@ func TestAsyncTerraformPluginSDKObserve(t *testing.T) {
 					},
 				},
 				cfg: cfgAsync,
-				obj: objAsync,
+				obj: newObjAsync(),
 			},
 			want: want{
 				obs: managed.ExternalObservation{
@@ -177,7 +163,7 @@ func TestAsyncTerraformPluginSDKObserve(t *testing.T) {
 					},
 				},
 				cfg: cfgAsync,
-				obj: objAsync,
+				obj: newObjAsync(),
 			},
 			want: want{
 				obs: managed.ExternalObservation{
@@ -226,7 +212,7 @@ func TestAsyncTerraformPluginSDKCreate(t *testing.T) {
 					},
 				},
 				cfg: cfgAsync,
-				obj: objAsync,
+				obj: newObjAsync(),
 				fns: CallbackFns{
 					CreateFn: func(nn types.NamespacedName) terraform.CallbackFn {
 						return func(err error, ctx context.Context) error {
@@ -242,7 +228,7 @@ func TestAsyncTerraformPluginSDKCreate(t *testing.T) {
 			terraformPluginSDKAsyncExternal := prepareTerraformPluginSDKAsyncExternal(tc.args.r, tc.args.cfg, tc.args.fns)
 			_, err := terraformPluginSDKAsyncExternal.Create(context.TODO(), tc.args.obj)
 			if diff := cmp.Diff(tc.want.err, err, test.EquateErrors()); diff != "" {
-				t.Errorf("\n%s\nConnect(...): -want error, +got error:\n", diff)
+				t.Errorf("\n%s\nterraformPluginSDKAsyncExternal.Create(...): -want error, +got error:\n", diff)
 			}
 		})
 	}
@@ -270,7 +256,7 @@ func TestAsyncTerraformPluginSDKUpdate(t *testing.T) {
 					},
 				},
 				cfg: cfgAsync,
-				obj: objAsync,
+				obj: newObjAsync(),
 				fns: CallbackFns{
 					UpdateFn: func(nn types.NamespacedName) terraform.CallbackFn {
 						return func(err error, ctx context.Context) error {
@@ -286,7 +272,7 @@ func TestAsyncTerraformPluginSDKUpdate(t *testing.T) {
 			terraformPluginSDKAsyncExternal := prepareTerraformPluginSDKAsyncExternal(tc.args.r, tc.args.cfg, tc.args.fns)
 			_, err := terraformPluginSDKAsyncExternal.Update(context.TODO(), tc.args.obj)
 			if diff := cmp.Diff(tc.want.err, err, test.EquateErrors()); diff != "" {
-				t.Errorf("\n%s\nConnect(...): -want error, +got error:\n", diff)
+				t.Errorf("\n%s\nterraformPluginSDKAsyncExternal.Update(...): -want error, +got error:\n", diff)
 			}
 		})
 	}
@@ -314,7 +300,7 @@ func TestAsyncTerraformPluginSDKDelete(t *testing.T) {
 					},
 				},
 				cfg: cfgAsync,
-				obj: objAsync,
+				obj: newObjAsync(),
 				fns: CallbackFns{
 					DestroyFn: func(nn types.NamespacedName) terraform.CallbackFn {
 						return func(err error, ctx context.Context) error {
@@ -330,8 +316,25 @@ func TestAsyncTerraformPluginSDKDelete(t *testing.T) {
 			terraformPluginSDKAsyncExternal := prepareTerraformPluginSDKAsyncExternal(tc.args.r, tc.args.cfg, tc.args.fns)
 			_, err := terraformPluginSDKAsyncExternal.Delete(context.TODO(), tc.args.obj)
 			if diff := cmp.Diff(tc.want.err, err, test.EquateErrors()); diff != "" {
-				t.Errorf("\n%s\nConnect(...): -want error, +got error:\n", diff)
+				t.Errorf("\n%s\nterraformPluginSDKAsyncExternal.Delete(...): -want error, +got error:\n", diff)
 			}
 		})
+	}
+}
+
+func newObjAsync() *fake.Terraformed {
+	return &fake.Terraformed{
+		Parameterizable: fake.Parameterizable{
+			Parameters: map[string]any{
+				"name": "example",
+				"map": map[string]any{
+					"key": "value",
+				},
+				"list": []any{"elem1", "elem2"},
+			},
+		},
+		Observable: fake.Observable{
+			Observation: map[string]any{},
+		},
 	}
 }

